### ISSUE TITLE
`slide_index_mean()` with window that reaches beyond `9999-12-31` errs

### ISSUE THREAD

**Issue by lschneiderbauer** [r-lib/slider#207]
``` r
vctrs::allow_lossy_cast(
  slider::slide_index_mean(
    c(1,2,3), c(as.Date("2022-12-31"), as.Date("2060-12-31"), as.Date("9999-12-31")),
    before = lubridate::duration("1 month"),
    after = lubridate::duration("1 month"),
    na_rm = TRUE
  )
)
#> Error in as.POSIXlt.character(x, tz, ...): Zeichenkette ist nicht in einem eindeutigen Standardformat
```

<sup>Created on 2024-07-29 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Although `as.Date("9999-12-31") + lubridate::duration("1 month")` should be a perfectly well defined date:
``` r
as.Date("9999-12-31") + lubridate::duration("1 month")
#> [1] "10000-01-30 10:30:00 UTC"
```

<sup>Created on 2024-07-29 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by DavisVaughan** [comment]
I have a few questions about this

If you _really_ do want `duration("1 month")`, which adds an _average month_ (i.e. average in number of seconds), then you should go ahead and convert your input to POSIXct (date-time) up front. Right now it is Date, going to POSIXct when it adds `duration("1 month")`, and then being forced back to Date (incorrectly, i think), and that is where the problem occurs. It's a vctrs casting problem more than anything, but supporting dates > 10000 years is out of scope for us, I think.

But anyways, if you just stay in POSIXct to begin with, it works

``` r
library(lubridate)

i <- c(as.Date("2022-12-31"), as.Date("2060-12-31"), as.Date("9999-12-31"))
i <- as.POSIXct(i, tz = "UTC")

slider::slide_index_mean(
  c(1,2,3), 
  i,
  before = lubridate::duration("1 month"),
  after = lubridate::duration("1 month"),
  na_rm = TRUE
)
#> [1] 1 2 3
```

Alternatively you might really want `months(1)` instead, i.e. a _period_ of 1 month not a _duration_ of 1 month. In that case, you could use `months()` combined with lubridate's "add with rollback" syntax to generate your endpoints and that would stay in Date the whole time.

``` r
library(lubridate)

i <- c(as.Date("2022-12-31"), as.Date("2060-12-31"), as.Date("9999-12-31"))

slider::slide_index_mean(
  c(1,2,3), 
  i,
  before = ~ .x %m-% months(1),
  after = ~ .x %m+% months(1),
  na_rm = TRUE
)
#> [1] 1 2 3

i %m-% months(1)
#> [1] "2022-11-30" "2060-11-30" "9999-11-30"
i %m+% months(1)
#> [1] "2023-01-31"  "2061-01-31"  "10000-01-31"
```

I think either of these is a much nicer way to solve your problem, so I don't think there is anything for slider to do, but thanks for the question!

**Comment by lschneiderbauer** [comment]
Thanks for the quick answer and suggestions!
I didn't realize that `slide_index_mean()` tries to cast back to the original type, that is also where the misplaced `vctrs::allow_lossy_cast()` is coming from.

In my use case the window size should really be of fixed length in physical units (and it is my understanding that the size of a month() in physical time units would vary).

Also, I want want to specify the full window size that is _centralized_ around the point of interest. So in this example I would want half a month in "before", and half a month in "after". But I can't divide a period by two, and so I am also led to use a "duration" instead of a "period".

To sum up, your first solution seems great for me! Thanks again!

**Comment by DavisVaughan** [comment]
Yea, definitely does sound like a case where you want a duration then! In that case, yea definitely suggest changing to the second precision POSIXct type ahead of time, since that is what durations are ultimately measured in.

I've also opened https://github.com/r-lib/slider/issues/208 to mention that `.before` and `.after` are cast back to the type of `.i`, since it does seem like I don't talk about that.
